bgcval2/functions/applyLandMask.py

```diff
@@ -0,0 +1,153 @@
+# Plymouth, PL1 3DH, UK
+#
+# Email:
+# ledm@pml.ac.uk
+#
+"""
+.. module:: applyLandMask
+   :platform: Unix
+   :synopsis: This function applies a land mask. Note that this will not work for p2p code.
+
+.. moduleauthor:: Lee de Mora <ledm@pml.ac.uk>
+
+"""
+
+import numpy as np
+from bgcval2.bgcvaltools.dataset import dataset
+
+
+tmask = {}
+
+
+def loadDataMask(areafile, maskname):
+    global tmask
+    nc = dataset(areafile, 'r')        
+    tmask[(areafile, maskname)] = nc.variables[maskname][:].squeeze()
+    nc.close()
+
+
+def applyLandMask(nc, keys, **kwargs):
+    try:
+        areafile = kwargs['areafile']
+    except:
+        raise AssertionError("applyLandMask:\t Needs an `areafile` kwarg to apply mask")    
+
+    if isinstance(areafile, list) and len(areafile)==1:
+        areafile = areafile[0]
+    else:
+        raise FileNotFoundError(f'Unable to find file: {areafile}')
+ 
+    try:
+        maskname = kwargs['maskname']
+    except:
+        maskname = 'tmask'
+
+    try:
+        mask = tmask[(areafile, maskname)]
+    except:
+        loadDataMask(areafile, maskname)
+        mask = tmask[(areafile, maskname)]
+    
+    try:
+        shape = mask.shape
+    except:
+        raise AssertionError("applyLandMask.py:\t Model mask not loaded correctly")
+
+    arr = np.ma.array(nc.variables[keys[0]][:]).squeeze()
+    m   = np.ma.masked_where(mask + arr.mask, arr)
+    m  += np.ma.masked_invalid(arr).mask
+    arr = np.ma.masked_where(m>1, arr)
+    print("applyLandMask:\t ", arr.mean(), arr.min(), arr.max(), arr.shape, m.mean())
+    assert 0
+    return arr
+
+
+def applyLandMask2D(nc, keys, **kwargs):
+    """
+    Useful for when the mask is 3D, but the field is only 2D,
+    so you want to apply the surface layer of the mask.
+    """
+    try:
+        areafile = kwargs['areafile']
+    except:
+        raise AssertionError("applyLandMask:\t Needs an `areafile` kwarg to apply mask")    
+   
+    if isinstance(areafile, list) and len(areafile)==1:
+        areafile = areafile[0]
+    else:
+        raise FileNotFoundError(f'Unable to find file: {areafile}')
+ 
+    try:
+        maskname = kwargs['maskname']
+    except:
+        maskname = 'tmask'
+
+    try:
+        mask = tmask[(areafile, maskname)][0]
+    except:
+        loadDataMask(areafile, maskname)
+        mask = tmask[(areafile, maskname)][0]
+            
+    try:
+        shape = mask.shape
+    except:
+        raise AssertionError("applyLandMask.py:\t Model mask not loaded correctly")
+
+    for key in keys:
+        if key not in nc.variables.keys():
+            print(f'key {key} not in file {nc.filename}')
+            print('Available keys:options:', nc.variables.keys())
+            raise KeyError(f'applyLandMask: key {key} not in file {nc.filename}.')
+    arr = np.ma.array(nc.variables[keys[0]][:]).squeeze()
+    m  = np.ma.masked_where(mask + arr.mask, arr)
+    m += np.ma.masked_invalid(arr).mask
+    return np.ma.masked_where(m, arr)
```

Comment:
this is almost identical to the above func, can you compress and use one func only?

Reply:
removed